common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp

```diff
@@ -139,10 +139,309 @@ static bool dwpal_get_channel_scan_freq(const std::vector<unsigned int> &channel
+        ie += length + 2;
+    }
+}
+
+static bool translate_nl_data_to_bwl_results(sChannelScanResults &results,
+                                             const struct nlattr **bss)
+{
+    //get bssid
+    std::copy_n(results.bssid.oct, sizeof(results.bssid.oct),
+                (unsigned char *)nla_data(bss[NL80211_BSS_BSSID]));
+
+    //get channel and operating frequency band
+    if (bss[NL80211_BSS_FREQUENCY]) {
+        int freq = nla_get_u32(bss[NL80211_BSS_FREQUENCY]);
+        if (freq >= 5180) {
+            results.operating_frequency_band = eOperating_Freq_Band_5GHz;
+        } else {
+            results.operating_frequency_band = eOperating_Freq_Band_2_4GHz;
+        }
+        results.channel = beerocks::utils::wifi_freq_to_channel(freq);
+    }
+
+    // get beacon period
+    if (bss[NL80211_BSS_BEACON_INTERVAL]) {
+        results.beacon_period_ms = (unsigned int)nla_get_u16(bss[NL80211_BSS_BEACON_INTERVAL]);
+    }
+
+    // get signal strength
+    if (bss[NL80211_BSS_SIGNAL_UNSPEC]) {
+        results.signal_strength_dBm = (nla_get_u8(bss[NL80211_BSS_SIGNAL_UNSPEC])) / 100;
+    } else if (bss[NL80211_BSS_SIGNAL_MBM]) {
+        results.signal_strength_dBm = (nla_get_u32(bss[NL80211_BSS_SIGNAL_MBM])) / 100;
+    }
+
+    //get information elements from information-elements-buffer or from beacon
+    if (bss[NL80211_BSS_BEACON_IES]) {
+        enum nl80211_bss ies_index = (bss[NL80211_BSS_INFORMATION_ELEMENTS])
+                                         ? NL80211_BSS_INFORMATION_ELEMENTS
+                                         : NL80211_BSS_BEACON_IES;
+        parse_info_elements((unsigned char *)nla_data(bss[ies_index]), nla_len(bss[ies_index]),
+                            results);
+    }
+
+    //get capabilities: mode, security_mode_enabled
+    if (bss[NL80211_BSS_CAPABILITY]) {
+        __u16 capa = nla_get_u16(bss[NL80211_BSS_CAPABILITY]);
+
+        if (capa & WLAN_CAPABILITY_IBSS) {
+            results.mode = eMode_AdHoc;
+        } else if (capa & WLAN_CAPABILITY_ESS) {
+            results.mode = eMode_Infrastructure;
+        }
+
+        if (results.security_mode_enabled.size() == 0) {
+            if (capa & WLAN_CAPABILITY_PRIVACY) {
+                results.security_mode_enabled.push_back(eSecurity_Mode_WEP);
+            } else {
+                results.security_mode_enabled.push_back(eSecurity_Mode_None);
+            }
+        }
+    }
+
+    return true;
+}
+
 static bool get_scan_results_from_nl_msg(sChannelScanResults &results, struct nl_msg *msg)
 {
-    LOG(ERROR) << "not suppoted";
-    return false;
+    struct nlattr *bss[NL80211_BSS_MAX + 1];
+
+    if (msg == nullptr) {
+        LOG(ERROR) << "invalid input: msg==NULL" << msg;
+        return false;
+    }
+
+    //prepare
+    memset(&results, '\0', sizeof(sChannelScanResults));
+
+    //read msg buffer into nl attributes struct
+    if (!read_nl_data_from_msg(bss, msg)) {
+        LOG(ERROR) << "failed to read nl data from msg";
```

Comment:
Not a very informative message. What does it actually means?
Netlink message failed to parse?